application/apps/indexer/addons/bufread/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+        self.slice.len()
+    }
+
+    /// Clears the buffer by consuming all available bytes
+    /// and returns the number of bytes being cleared.
+    pub fn clear(&mut self) -> usize {
+        let size = self.read_available();
+        self.read_done(size)
+    }
+
+    /// Reads from this buffer into the given output
+    /// and returns the number of bytes being read.
+    pub fn read_to(&mut self, buffer: &mut [u8]) -> usize {
+        let size = min(self.read_available(), buffer.len());
+        buffer[..size].copy_from_slice(&self.read_slice()[..size]);
+        self.read_done(size)
+    }
+
+    /// Returns the number of currently available bytes for reading.
+    pub fn read_available(&self) -> usize {
+        self.end - self.start
+    }
+
+    /// Returns the current slice to read from.
+    pub fn read_slice(&self) -> &[u8] {
+        &self.slice[self.start..self.end]
+    }
+
+    /// Signals the amount of newly read bytes from the buffer
+    /// and returns the number of bytes being updated.
+    pub fn read_done(&mut self, size: usize) -> usize {
+        let before = self.start;
+
+        self.start = min(self.start + size, self.slice.len());
+        let diff = self.start - before;
+
+        if self.start == self.end {
+            self.start = 0;
+            self.end = 0;
+        }
+
+        diff
+    }
+
+    /// Writes from the given input into this buffer
+    /// and returns the number of bytes being written.
+    pub fn write_from(&mut self, buffer: &[u8]) -> usize {
+        let size = min(self.write_available(), buffer.len());
+        self.write_slice()[..size].copy_from_slice(&buffer[..size]);
+        self.write_done(size)
+    }
+
+    /// Returns the number of currently available bytes for writing.
+    pub fn write_available(&self) -> usize {
+        self.slice.len() - self.end
+    }
+
+    /// Returns the current slice to write to.
+    pub fn write_slice(&mut self) -> &mut [u8] {
+        &mut self.slice[self.end..]
+    }
+
+    /// Signals the amount of newly written bytes to the buffer
+    /// and returns the number of bytes being updated.
+    pub fn write_done(&mut self, size: usize) -> usize {
+        let before = self.end;
+
+        self.end = min(self.end + size, self.slice.len());
+        self.end - before
+    }
+
+    /// Moves any remaining bytes within the buffer to its front
+    /// and returns the number of bytes being moved.
+    pub fn flush(&mut self) -> usize {
+        let before = self.start;
+
+        if self.start != 0 {
+            // SAFETY: Copying the bytes is safe as pointers will not overlap.
+            unsafe {
+                copy(
+                    self.read_slice().as_ptr(),
+                    self.slice.as_mut_ptr(),
+                    self.read_available(),
+                );
+            }
+
+            self.end -= self.start;
+            self.start = 0;
+        }
+
+        before
+    }
+}
+
+#[cfg(test)]
```

Comment:
since we have a `tests` module, why not put all the tests in it?

Reply:
This are local unit-tests for the code in this source file and I would keep them here. In the tests module there are the prob-tests and some more comprehensive ("integration") tests using the crate for a pseudo-protocol parser, however.